pkg/skaffold/deploy/docker/port.go

```diff
@@ -0,0 +1,139 @@
+)
+
+type PortManager struct {
+	containerPorts map[string][]int // maps containers to the ports they have allocated
+	portSet        util.PortSet
+
+	lock sync.Mutex
+}
+
+func NewPortManager() *PortManager {
+	return &PortManager{
+		containerPorts: make(map[string][]int),
+		portSet:        util.PortSet{},
+	}
+}
+
+// getPorts converts PortForwardResources into docker.PortSet/PortMap objects.
+// These are passed to ContainerCreate on Deploy to expose container ports on the host.
+// It also returns a list of containerPortForwardEntry, to be passed to the event handler
+func (pm *PortManager) getPorts(containerName string, pf []*v1.PortForwardResource) (nat.PortSet, nat.PortMap, []containerPortForwardEntry, error) {
+	pm.lock.Lock()
+	defer pm.lock.Unlock()
+	s := make(nat.PortSet)
+	m := make(nat.PortMap)
+	var entries []containerPortForwardEntry
+	var ports []int
+	for _, p := range pf {
+		if strings.ToLower(string(p.Type)) != "container" {
+			logrus.Debugf("skipping non-container port forward resource in Docker deploy: %s\n", p.Name)
+			continue
+		}
+		localPort := util.GetAvailablePort(p.Address, p.LocalPort, &pm.portSet)
+		ports = append(ports, localPort)
+		port, err := nat.NewPort("tcp", p.Port.String())
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		s[port] = struct{}{}
+		m[port] = []nat.PortBinding{
+			{HostIP: p.Address, HostPort: fmt.Sprintf("%d", localPort)},
+		}
+		entries = append(entries, containerPortForwardEntry{
+			container:       containerName,
+			resourceName:    p.Name,
+			resourceAddress: p.Address,
+			localPort:       int32(localPort),
+			remotePort:      p.Port,
+		})
+	}
+	pm.containerPorts[containerName] = ports
+	return s, m, entries, nil
+}
+
+func (pm *PortManager) relinquishPorts(containerName string) {
+	pm.lock.Lock()
+	defer pm.lock.Unlock()
+	ports := pm.containerPorts[containerName]
+	for _, port := range ports {
+		pm.portSet.Delete(port)
+	}
+	pm.containerPorts[containerName] = nil
+}
+
+type containerPortForwardEntry struct {
+	container       string
+	resourceName    string
+	resourceAddress string
+	localPort       int32
+	remotePort      schemautil.IntOrString
+}
+
+func containerPortForwardEvents(out io.Writer, entries []containerPortForwardEntry) {
+	for _, entry := range entries {
+		event.PortForwarded(
+			entry.localPort,
+			entry.remotePort,
+			"",              // no pod name
+			entry.container, // container name
+			"",              // no namespace
+			"",              // no port name
+			"container",
+			entry.resourceName,
+			entry.resourceAddress,
+		)
+
+		eventV2.PortForwarded(
+			entry.localPort,
+			entry.remotePort,
+			"",              // no pod name
+			entry.container, // container name
+			"",              // no namespace
+			"",              // no port name
+			"container",
+			entry.resourceName,
+			entry.resourceAddress,
+		)
+
+		output.Green.Fprintln(out,
+			fmt.Sprintf("[%s] Forwarding container port %s -> local port %s:%d",
```

Comment:
nit: prefix this with `http://` to make the link clickable
```suggestion
			fmt.Sprintf("[%s] Forwarding container port %s -> local port http://%s:%d",
```

Reply:
Also wondering if this output should still just happen in the port forward phase or accessor logic just so things are properly categorized for CC ides. Having these lines printed here makes their corresponding events come labelled as being from `Deploy` when we would want them as `PortForward`. Not sure if this would be a huge pain on our end

Reply:
yeah this is a good point. updated so that the `PortManager` fills the `Accessor` interface, and the `Start()` method fires the events and the terminal output.